File: java17/src/main/java/folio/codinginterview/presentation/PresentationPreparation.java
package folio.codinginterview.presentation;

import folio.codinginterview.domain.UserId;
import folio.codinginterview.presentation.PresentationException.BadRequestException;

import java.math.BigDecimal;
import java.util.concurrent.CompletableFuture;

public abstract class PresentationPreparation {
    protected CompletableFuture<UserId> parseUserId(String s) {
        try {
            return CompletableFuture.completedFuture(new UserId(s));
        } catch (RuntimeException e) {
            CompletableFuture<UserId> failed = new CompletableFuture<>();
            failed.completeExceptionally(new BadRequestException(e.getMessage()));
            return failed;
        }
    }

    protected CompletableFuture<BigDecimal> parseAmount(String s) {
        try {
            return CompletableFuture.completedFuture(new BigDecimal(s));
        } catch (RuntimeException e) {
            CompletableFuture<BigDecimal> failed = new CompletableFuture<>();
            failed.completeExceptionally(new BadRequestException("invalid amount: " + s));
            return failed;
        }
    }
}

File: ruby/lib/coding_interview/presentation/presentation_preparation.rb
require "bigdecimal"
require_relative "presentation_exception"
require_relative "../domain/user_id"

module CodingInterview
  module Presentation
    module PresentationPreparation
      def parse_user_id(s)
        Domain::UserId.new(s)
      rescue ArgumentError => e
        raise BadRequestException.new(e.message)
      end

      def parse_amount(s)
        BigDecimal(s)
      rescue ArgumentError, TypeError
        raise BadRequestException.new("invalid amount: #{s}")
      end
    end
  end
end
